File: server/core/service/api_service.go
package service

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/nexus-flow/nexus/core/model"
	"github.com/nexus-flow/nexus/core/repository"
)

type APIInfoDTO struct {
	ID                    int64       `json:"id"`
	SuiteID               int64       `json:"suiteId"`
	SuiteFlag             *int        `json:"suiteFlag,omitempty"`
	ApiURL                string      `json:"apiUrl"`
	ApiName               string      `json:"apiName"`
	ApiDesc               string      `json:"apiDesc,omitempty"`
	ApiRequestType        string      `json:"apiRequestType"`
	ApiRequestContentType string      `json:"apiRequestContentType,omitempty"`
	ApiHeaders            interface{} `json:"apiHeaders,omitempty"`
	ApiInputParams        interface{} `json:"apiInputParams,omitempty"`
	ApiOutputParams       interface{} `json:"apiOutputParams,omitempty"`
}

func CreateAPI(a *model.API) error {
	return repository.CreateAPI(a)
}

func UpdateAPI(a *model.API) error {
	return repository.UpdateAPI(a)
}

func DeleteAPI(id int64) error {
	return repository.DeleteAPI(id)
}

func GetAPIInfo(id int64) (*APIInfoDTO, error) {
	a, err := repository.GetAPIByID(id)
	if err != nil {
		return nil, err
	}
	return buildAPIInfoDTO(a)
}

func GetAPIInfoByCode(code string) (*APIInfoDTO, error) {
	a, err := repository.GetAPIByCode(code)
	if err != nil {
		return nil, err
	}
	return buildAPIInfoDTO(a)
}

func buildAPIInfoDTO(a *model.API) (*APIInfoDTO, error) {
	dto := &APIInfoDTO{
		ID:                    a.ID,
		SuiteID:               a.SuiteID,
		ApiURL:                a.ApiURL,
		ApiName:               a.ApiName,
		ApiDesc:               a.ApiDesc,
		ApiRequestType:        a.ApiRequestType,
		ApiRequestContentType: a.ApiRequestContentType,
	}
	suite, err := repository.GetSuiteByID(a.SuiteID)
	if err == nil {
		dto.SuiteFlag = suite.SuiteFlag
	}

	// Load parameters
	headers, _ := repository.ListParametersBySourceAndType("api", a.ID, 4) // HEADER
	inputs, _ := repository.ListParametersBySourceAndType("api", a.ID, 1)  // INPUT
	outputs, _ := repository.ListParametersBySourceAndType("api", a.ID, 2) // OUTPUT
	dto.ApiHeaders = headers
	dto.ApiInputParams = inputs
	dto.ApiOutputParams = outputs
	return dto, nil
}

func ListAPIsBySuiteID(suiteID int64) ([]model.API, error) {
	return repository.ListAPIsBySuiteID(suiteID)
}

func ListAPIsBySuiteCode(suiteCode string) ([]model.API, error) {
	return repository.ListAPIsBySuiteCode(suiteCode)
}

func PageAPIs(p *model.PageParam, suiteID *int64, apiName, apiURL string) ([]model.API, int64, error) {
	return repository.PageAPIs(p, suiteID, apiName, apiURL)
}

func DebugAPI(apiID int64, headerData, inputParamData map[string]interface{}) (map[string]interface{}, error) {
	a, err := repository.GetAPIByID(apiID)
	if err != nil {
		return nil, fmt.Errorf("接口不存在")
	}

	url := a.ApiURL
	// Replace path variables
	for k, v := range inputParamData {
		placeholder := "{" + k + "}"
		if strings.Contains(url, placeholder) {
			url = strings.ReplaceAll(url, placeholder, fmt.Sprintf("%v", v))
			delete(inputParamData, k)
		}
	}

	client := &http.Client{Timeout: 30 * time.Second}
	var req *http.Request

	switch strings.ToUpper(a.ApiRequestType) {
	case "GET":
		req, _ = http.NewRequest("GET", url, nil)
		q := req.URL.Query()
		for k, v := range inputParamData {
			q.Set(k, fmt.Sprintf("%v", v))
		}
		req.URL.RawQuery = q.Encode()
	default:
		body, _ := json.Marshal(inputParamData)
		req, _ = http.NewRequest(strings.ToUpper(a.ApiRequestType), url, strings.NewReader(string(body)))
		ct := a.ApiRequestContentType
		if ct == "" {
			ct = "application/json"
		}
		req.Header.Set("Content-Type", ct)
	}

	for k, v := range headerData {
		req.Header.Set(k, fmt.Sprintf("%v", v))
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("请求失败: %v", err)
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	var result map[string]interface{}
	if json.Unmarshal(respBody, &result) != nil {
		result = map[string]interface{}{"response": string(respBody)}
	}
	return result, nil
}

func SaveAPIParameters(apiID int64, sourceType string, headers, inputs, outputs []model.Parameter) error {
	repository.DeleteParametersBySource(sourceType, apiID)

	for i := range headers {
		headers[i].SourceType = sourceType
		headers[i].SourceID = &apiID
		pt := 4
		headers[i].ParamType = &pt
	}
	for i := range inputs {
		inputs[i].SourceType = sourceType
		inputs[i].SourceID = &apiID
		pt := 1
		inputs[i].ParamType = &pt
	}
	for i := range outputs {
		outputs[i].SourceType = sourceType
		outputs[i].SourceID = &apiID
		pt := 2
		outputs[i].ParamType = &pt
	}

	all := append(append(headers, inputs...), outputs...)
	return repository.BatchCreateParameters(all)
}

File: server/src/main/java/com/javaoffers/nexus/core/service/ApiService.java
package com.javaoffers.nexus.core.service;

import com.javaoffers.nexus.core.engine.HttpExecutor;
import com.javaoffers.nexus.core.engine.MethodDef;
import com.javaoffers.nexus.core.model.API;
import com.javaoffers.nexus.core.model.PageParam;
import com.javaoffers.nexus.core.model.Parameter;
import com.javaoffers.nexus.core.model.Suite;
import com.javaoffers.nexus.core.repository.ApiMapper;
import com.javaoffers.nexus.core.repository.ParameterMapper;
import com.javaoffers.nexus.core.repository.SuiteMapper;
import com.javaoffers.nexus.pkg.exception.BizException;
import lombok.Data;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

/** 对应 Go core/service/api_service.go。 */
@Service
public class ApiService {

    private final ApiMapper apiMapper;
    private final SuiteMapper suiteMapper;
    private final ParameterMapper parameterMapper;
    private final HttpExecutor httpExecutor;

    public ApiService(ApiMapper apiMapper, SuiteMapper suiteMapper, ParameterMapper parameterMapper, HttpExecutor httpExecutor) {
        this.apiMapper = apiMapper;
        this.suiteMapper = suiteMapper;
        this.parameterMapper = parameterMapper;
        this.httpExecutor = httpExecutor;
    }

    @Data
    public static class APIInfoDTO {
        private Long id;
        private Long suiteId;
        private Integer suiteFlag;
        private String apiUrl;
        private String apiName;
        private String apiDesc;
        private String apiRequestType;
        private String apiRequestContentType;
        private List<Parameter> apiHeaders;
        private List<Parameter> apiInputParams;
        private List<Parameter> apiOutputParams;
    }

    public void createAPI(API a) {
        apiMapper.create(a);
        // brief 不回填自增 id，按 apiCode 取回真实 id（无 apiCode 时取最新行）
        API saved = (a.getApiCode() != null && !a.getApiCode().isEmpty())
                ? apiMapper.getByCode(a.getApiCode())
                : apiMapper.findNewest();
        if (saved != null) {
            a.setId(saved.getId());
        }
    }

    public void updateAPI(API a) {
        apiMapper.update(a);
    }

    public void deleteAPI(Long id) {
        apiMapper.delete(id);
    }

    public APIInfoDTO getAPIInfo(Long id) {
        API a = apiMapper.getById(id);
        if (a == null) {
            throw new BizException("接口不存在");
        }
        return buildDTO(a);
    }

    public APIInfoDTO getAPIInfoByCode(String code) {
        API a = apiMapper.getByCode(code);
        if (a == null) {
            throw new BizException("接口不存在");
        }
        return buildDTO(a);
    }

    private APIInfoDTO buildDTO(API a) {
        APIInfoDTO dto = new APIInfoDTO();
        dto.setId(a.getId());
        dto.setSuiteId(a.getSuiteId());
        dto.setApiUrl(a.getApiUrl());
        dto.setApiName(a.getApiName());
        dto.setApiDesc(a.getApiDesc());
        dto.setApiRequestType(a.getApiRequestType());
        dto.setApiRequestContentType(a.getApiRequestContentType());
        Suite suite = suiteMapper.getById(a.getSuiteId());
        if (suite != null) {
            dto.setSuiteFlag(suite.getSuiteFlag());
        }
        dto.setApiHeaders(parameterMapper.listBySourceAndType("api", a.getId(), 4));
        dto.setApiInputParams(parameterMapper.listBySourceAndType("api", a.getId(), 1));
        dto.setApiOutputParams(parameterMapper.listBySourceAndType("api", a.getId(), 2));
        return dto;
    }

    public List<API> listAPIsBySuiteID(Long suiteID) {
        return apiMapper.listBySuiteId(suiteID);
    }

    public List<API> listAPIsBySuiteCode(String suiteCode) {
        Suite suite = suiteMapper.getByCode(suiteCode);
        if (suite == null) {
            throw new BizException("套件不存在");
        }
        return apiMapper.listBySuiteId(suite.getId());
    }

    public Page<API> pageAPIs(PageParam p, Long suiteID, String apiName, String apiURL) {
        int pn = p.getPageNum() <= 0 ? 1 : p.getPageNum();
        int ps = p.getPageSize() <= 0 ? 10 : p.getPageSize();
        List<API> list = apiMapper.page(pn, ps, suiteID, apiName, apiURL);
        long total = apiMapper.count(suiteID, apiName, apiURL);
        return new Page<>(total, list);
    }

    public Map<String, Object> debugAPI(Long apiID, Map<String, Object> headerData, Map<String, Object> inputParamData) {
        API a = apiMapper.getById(apiID);
        if (a == null) {
            throw new BizException("接口不存在");
        }
        MethodDef method = new MethodDef();
        method.setUrl(a.getApiUrl());
        method.setRequestType(a.getApiRequestType());
        method.setRequestContentType(a.getApiRequestContentType());
        Map<String, Object> result = httpExecutor.sendRequest(method, headerData, inputParamData);
        if (result == null) {
            throw new BizException("请求失败");
        }
        return result;
    }

    public void saveAPIParameters(Long apiID, String sourceType,
                                  List<Parameter> headers, List<Parameter> inputs, List<Parameter> outputs) {
        parameterMapper.deleteBySource(sourceType, apiID);
        if (headers != null) {
            for (Parameter p : headers) {
                p.setSourceType(sourceType);
                p.setSourceId(apiID);
                p.setParamType(4);
            }
        }
        if (inputs != null) {
            for (Parameter p : inputs) {
                p.setSourceType(sourceType);
                p.setSourceId(apiID);
                p.setParamType(1);
            }
        }
        if (outputs != null) {
            for (Parameter p : outputs) {
                p.setSourceType(sourceType);
                p.setSourceId(apiID);
                p.setParamType(2);
            }
        }
        java.util.List<Parameter> all = new java.util.ArrayList<>();
        if (headers != null) all.addAll(headers);
        if (inputs != null) all.addAll(inputs);
        if (outputs != null) all.addAll(outputs);
        parameterMapper.batchCreate(all);
    }
}
